atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/AbstractAtlasDbKeyValueServiceTest.java

```diff
@@ -660,21 +661,104 @@ public void testGetRangeOfTimestampsThrowsOnError() {
 
     @Test
     public void testCannotModifyValuesAfterWrite() {
+        Cell cell = Cell.create(row0, column0);
         byte[] data = new byte[1];
-        byte[] unmodifiedData = Arrays.copyOf(data, data.length);
+        byte[] unmodifiedData = writeToCell(cell, data);
+
+        data[0] = (byte) 50;
+
+        assertThat(getForCell(cell), is(unmodifiedData));
+
+        keyValueService.delete(TEST_TABLE, ImmutableMultimap.of(cell, TEST_TIMESTAMP + 1));
+    }
 
+    @Test
+    public void testCannotModifyValuesAfterGetRows() {
         Cell cell = Cell.create(row0, column0);
+        byte[] unmodifiedData = writeToCell(cell, new byte[1]);
 
-        Value val = Value.create(data, TEST_TIMESTAMP + 1);
+        modifyReference(getRowsForCell(cell));
+
+        assertThat(getRowsForCell(cell), is(unmodifiedData));
+    }
+
+    @Test
+    public void testCannotModifyValuesAfterGet() {
+        Cell cell = Cell.create(row0, column0);
+        byte[] unmodifiedData = writeToCell(cell, new byte[1]);
+
+        modifyReference(getForCell(cell));
+
+        assertThat(getForCell(cell), is(unmodifiedData));
+    }
+
+    @Test
+    public void testCannotModifyValuesAfterGetRange() {
+        Cell cell = Cell.create(row0, column0);
+        byte[] unmodifiedData = writeToCell(cell, new byte[1]);
+
+        modifyReference(getOnlyItemInTableRange());
+
+        assertThat(getOnlyItemInTableRange(), is(unmodifiedData));
+    }
+
+    @Test
+    public void testCannotModifyValuesAfterGetRangeWithHistoryIfSupported() {
+        if (kvsSupportsGetRangeWithHistory()) {
```

Comment:
@schlosna Thanks! 

@joelea I haven't seen it elsewhere, but (especially combined with `assumeTrue`) it feels like an acceptable workaround to not all KVSs implementing the entire API. I considered making this test specific to InMemoryKVSTest, but that feels worse.